drivers/pn532/pn532.c

```diff
@@ -0,0 +1,639 @@
+
+int pn532_get_passive_iso14443a(pn532_t *dev, nfc_iso14443a_t *out,
+                                unsigned max_retries)
+{
+    int ret = -1;
+    char buff[PN532_BUFFER_LEN];
+
+    if (_set_act_retries(dev, buff, max_retries) == 0) {
+        ret = _list_passive_targets(dev, buff, PN532_BR_106_ISO_14443_A, 1,
+                                    LIST_PASSIVE_LEN_14443(1));
+    }
+
+    if (ret > 0 && buff[0] > 0) {
+        out->target = buff[1];
+        out->sns_res = (buff[2] << 8) | buff[3];
+        out->sel_res = buff[4];
+        out->id_len  = buff[5];
+        out->type = ISO14443A_UNKNOWN;
+
+        for (int i = 0; i < out->id_len; i++) {
+            out->id[i] = buff[6 + i];
+        }
+
+        /* try to find out the type */
+        if (out->id_len == 4) {
+            out->type = ISO14443A_MIFARE;
+
+        }
+        else if (out->id_len == 7) {
+            /* In the case of type 4, the first byte of RATS is the length
+             * of RATS including the length itself (6+7) */
+            if (buff[13] == ret - 13) {
+                out->type = ISO14443A_TYPE4;
+            }
+        }
+
+        ret = 0;
+
+    }
+    else {
+        ret = -1;
+    }
+
+    return ret;
+}
+
+void pn532_deselect_passive(pn532_t *dev, unsigned target_id)
+{
+    char buff[PN532_BUFFER_LEN];
+
+    buff[BUFF_CMD_START     ] = CMD_DESELECT;
+    buff[BUFF_DATA_START    ] = target_id;
+
+    send_rcv(dev, buff, 1, 1);
+
+}
+
+void pn532_release_passive(pn532_t *dev, unsigned target_id)
+{
+    char buff[PN532_BUFFER_LEN];
+
+    buff[BUFF_CMD_START     ] = CMD_RELEASE;
+    buff[BUFF_DATA_START    ] = target_id;
+
+    send_rcv(dev, buff, 1, 1);
+
+}
+
+int pn532_mifareclassic_authenticate(pn532_t *dev, nfc_iso14443a_t *card,
+                                     pn532_mifare_key_t keyid, char *key, unsigned block)
+{
+    int ret = -1;
+    char buff[PN532_BUFFER_LEN];
+
+    buff[BUFF_CMD_START     ] = CMD_DATA_EXCHANGE;
+    buff[BUFF_DATA_START    ] = card->target;
+    buff[BUFF_DATA_START + 1] = keyid;
+    buff[BUFF_DATA_START + 2] = block; /* current block */
+
+    /* key */
+    for (int i = 0; i < 6; i++) {
+        buff[BUFF_DATA_START + 3 + i] = key[i];
+    }
+
+    /* id */
+    for (int i = 0; i < card->id_len; i++) {
+        buff[BUFF_DATA_START + 9 + i] = card->id[i];
+    }
+
+    ret = send_rcv(dev, buff, 9 + card->id_len, 1);
+    if (ret == 1) {
+        ret = buff[0];
+        card->auth = 1;
+    }
+
+    return ret;
+}
+
+int pn532_mifareclassic_write(pn532_t *dev, char *idata, nfc_iso14443a_t *card,
+                              unsigned block)
+{
+    int ret = -1;
+    char buff[PN532_BUFFER_LEN];
+
+    if (card->auth) {
+
+        buff[BUFF_CMD_START     ] = CMD_DATA_EXCHANGE;
+        buff[BUFF_DATA_START    ] = card->target;
+        buff[BUFF_DATA_START + 1] = MIFARE_CMD_WRITE;
+        buff[BUFF_DATA_START + 2] = block; /* current block */
+        memcpy(&buff[BUFF_DATA_START - 1 + 4], idata, 16);
```

Comment:
Mifare can read chunks of 16 bytes at a time only (if I remember correctly). I'll add a constant for that :+1: